File: backend/app/report.py
from typing import List, Optional
from backend.app.models import Finding, DecompileStats, ManifestData, Report, ReportSummary
from backend.app.scoring import compute_score, score_to_grade, compute_summary, sort_findings

def generate_report(
    app_name: str,
    findings: List[Finding],
    decompilation_stats: Optional[DecompileStats] = None,
    manifest_data: Optional[ManifestData] = None
) -> Report:
    """
    Generates structured security assessment report matching the single source of truth schema.
    """
    resolved_app_name = app_name
    if manifest_data and manifest_data.package_name:
        resolved_app_name = manifest_data.package_name

    score = compute_score(findings)
    grade = score_to_grade(score)
    summary = compute_summary(findings)
    sorted_findings = sort_findings(findings)

    decomp_incomplete = False
    decomp_warnings: List[str] = []

    if decompilation_stats:
        decomp_incomplete = decompilation_stats.decompilation_incomplete or decompilation_stats.status in ("failed", "partial")
        decomp_warnings = list(decompilation_stats.decompilation_warnings)
        if decompilation_stats.error and decompilation_stats.error not in decomp_warnings:
            decomp_warnings.append(decompilation_stats.error)

    return Report(
        app_name=resolved_app_name,
        score=score,
        grade=grade,
        decompilation_incomplete=decomp_incomplete,
        decompilation_warnings=decomp_warnings,
        findings=sorted_findings,
        summary=summary
    )

File: src/lib/analyzer/report.ts
import { DecompileStats, Finding, ManifestData, Report } from "./types";
import { computeScore, scoreToGrade, computeSummary, sortFindings } from "./scoring";

export function generateReport(
  appName: string,
  findings: Finding[],
  decompilationStats?: DecompileStats | null,
  manifestData?: ManifestData | null
): Report {
  const resolvedAppName = manifestData?.package_name || appName;
  const score = computeScore(findings);
  const grade = scoreToGrade(score);
  const summary = computeSummary(findings);
  const sortedFindings = sortFindings(findings);

  let decompIncomplete = false;
  let decompWarnings: string[] = [];

  if (decompilationStats) {
    decompIncomplete =
      decompilationStats.decompilation_incomplete ||
      decompilationStats.status === "failed" ||
      decompilationStats.status === "partial";
    decompWarnings = [...decompilationStats.decompilation_warnings];
    if (decompilationStats.error && !decompWarnings.includes(decompilationStats.error)) {
      decompWarnings.push(decompilationStats.error);
    }
  }

  return {
    app_name: resolvedAppName,
    score,
    grade,
    decompilation_incomplete: decompIncomplete,
    decompilation_warnings: decompWarnings,
    findings: sortedFindings,
    summary,
  };
}
